src/runtime/virtcontainers/clh.go

```diff
@@ -1347,9 +1347,22 @@ func (clh *cloudHypervisor) bootVM(ctx context.Context) error {
 		return fmt.Errorf("VM state is not 'Created' after 'CreateVM'")
 	}
 
-	err = clh.vmAddNetPut()
-	if err != nil {
-		return err
+	for {
+		err = clh.vmAddNetPut()
+		if err != nil {
+			return err
+		}
+
+		info, err = clh.vmInfo()
+		if err != nil {
+			return err
+		}
+
+		vmConfig := info.GetConfig()
+		netConfig := vmConfig.GetNet()
+		if len(netConfig) == len(*clh.netDevices) {
+			break
+		}
```

Comment:
This is really a workaround and I'd prefer if we could find the root cause of the problem rather than merging this.
I was thinking, maybe the socket is opened as nonblocking, meaning right after the message is written [here](https://github.com/fidencio/kata-containers/blob/2fca77ebbeaccb6dd81666183aca7158da97ddd3/src/runtime/virtcontainers/clh.go#L188) the connection is closed because of the [defer statement](https://github.com/fidencio/kata-containers/blob/2fca77ebbeaccb6dd81666183aca7158da97ddd3/src/runtime/virtcontainers/clh.go#L159), preventing the message from being delivered.
You can try several things to identify if this is related to the connection being closed too early:
- remove the `defer conn.Close()`
- add a sleep right after `conn.WriteMsgUnix([]byte(payload), oob, nil)`
- find a way to make the socket block until the message has been received

Reply:
Yeh, this is definitely wrong. I think that the lack of flush, the failure to check the return value of `sendmsg` and the absence of a read on the socket to get the response are the problem.

Reply:
Ah yes I agree with @rbradford there's an important aspect of the communication that's not covered here, the HTTP response. Waiting for the response will force you to read (hence the flush of the write will have been done already) the reply, which is the correct way of knowing that things went well. And you can get useful information from the reply (the device identifier and its PCI b/d/f).